org.osgi.test.assertj/src/main/java/org/osgi/test/assertj/bundle/AbstractBundleAssert.java

```diff
@@ -0,0 +1,366 @@
+				actual, expected, actual.getVersion());
+		}
+		return myself;
+	}
+
+	public SELF hasLastModified(Object expected) {
+		isNotNull();
+		long expectedTime = getTime(expected);
+		if (actual.getLastModified() != expectedTime) {
+			failWithMessage("%nExpecting%n <%s>%nto have last modified time:%n  <%d>%n but it was:%n  <%d>", actual,
+				expectedTime, actual.getLastModified());
+		}
+		return myself;
+	}
+
+	public SELF isModifiedAfter(Object expected) {
+		isNotNull();
+		long expectedTime = getTime(expected);
+		if (actual.getLastModified() <= expectedTime) {
+			failWithMessage("%nExpecting%n  <%s>%nto have been modified after:%n  <%d>%n but was modified at:%n  <%d>",
+				actual, expectedTime, actual.getLastModified());
+		}
+		return myself;
+	}
+
+	public SELF isModifiedAfterOrAt(Object expected) {
+		isNotNull();
+		long expectedTime = getTime(expected);
+		if (actual.getLastModified() < expectedTime) {
+			failWithMessage(
+				"%nExpecting%n  <%s>%nto have been modified after or at:%n  <%d>%n but was modified at:%n  <%d>",
+				actual, expectedTime, actual.getLastModified());
+		}
+		return myself;
+	}
+
+	public SELF isModifiedBefore(Object expected) {
+		isNotNull();
+		long expectedTime = getTime(expected);
+		if (actual.getLastModified() >= expectedTime) {
+			failWithMessage("%nExpecting%n  <%s>%nto have been modified before:%n  <%d>%n but was modified at:%n  <%d>",
+				actual, expectedTime, actual.getLastModified());
+		}
+		return myself;
+	}
+
+	public SELF isModifiedBeforeOrAt(Object expected) {
+		isNotNull();
+		long expectedTime = getTime(expected);
+		if (actual.getLastModified() > expectedTime) {
+			failWithMessage(
+				"%nExpecting%n  <%s>%nto have been modified before or at:%n  <%d>%n but was modified at:%n  <%d>",
+				actual, expectedTime, actual.getLastModified());
+		}
+		return myself;
+	}
+
+	private static long getTime(Object expected) {
+		if (expected == null) {
+			throw new IllegalArgumentException("expected cannot be null");
+		} else if (expected instanceof Long) {
+			return (Long) expected;
+		} else if (expected instanceof Date) {
+			return ((Date) expected).getTime();
+		} else if (expected instanceof Instant) {
+			return ((Instant) expected).toEpochMilli();
+		} else {
+			throw new IllegalArgumentException("Expected must be a long, Date or Instant");
+		}
+	}
+
+	public SELF isInState(int expected) {
+		isNotNull();
+		if (actual.getState() != expected) {
+			failWithMessage("%nExpecting%n  <%s>%nto be in state:%n  <%s>%n but was in state:%n  <%s>", actual,
+				stateToString(expected), stateToString(actual.getState()));
+		}
+		return myself;
+	}
+
+	public SELF isNotInState(int expected) {
+		isNotNull();
+		final String expectedState = stateToString(expected);
+		if (actual.getState() == expected) {
+			failWithMessage("%nExpecting%n  <%s>%nnot to be in state:%n  <%s>%n but it was", actual,
+				expectedState);
+		}
+		return myself;
+	}
+
+	public SELF isInStateMask(int mask) {
```

Comment:
The intended difference: 

`isInState()` tests only for a single state. If passed anything other than a single state, it will throw IAE if it is passed anything else (ie`, isInState(INSTALLED | ACTIVE)` will throw IAE). The idea was to guard against accidental errors in your test.

`isInStateMask()` allows a bit mask of state values; ie `isInStateMask(INSTALLED | ACTIVE)` is allowed.

I'm open to suggestions here if you think this syntax will be confusing for users.

Reply:
Even for a single state test, the Bundle.getState() method should be tested with an AND operation. While the OSGi spec does not currently ever set more than one bit (and probably can never do that ever since so many people test state with `==` and `!=`), the state is a bit mask and theoretically multiple bits could be set. So, as exemplars, we should never test state with `==` and `!=`. I see your point about multiple states passed to `isInState` but the method should check that the state argument has only a single bit set

`(state != 0) && ((state & (state - 1)) == 0)`

and then do a bit test against getState().

Reply:
Done.